mediators/datamapper/components/org.wso2.carbon.gateway.mediators.datamapper.engine/src/main/java/org/wso2/carbon/gateway/mediators/datamapper/engine/core/schemas/JacksonJSONSchema.java

```diff
@@ -0,0 +1,375 @@
+        try {
+            jsonSchemaMap = objectMapper.readValue(inputSchema, Map.class);
+        } catch (IOException e) {
+            throw new SchemaException("Error while reading input stream. " + e.getMessage());
+        }
+        initNamespaceMap();
+    }
+
+    /**
+     * populating name-space value map.
+     */
+    private void initNamespaceMap() {
+        namespaceMap = new HashMap<>();
+        ArrayList<Map> namespaceElementArray = (ArrayList<Map>) jsonSchemaMap.get(NAMESPACE_KEY);
+        if (namespaceElementArray != null) {
+            for (Map namespaceObject : namespaceElementArray) {
+                String urlValue = (String) namespaceObject.get(URL_KEY);
+                if (!namespaceMap.containsKey(urlValue)) {
+                    namespaceMap.put(urlValue, (String) namespaceObject.get(PREFIX_KEY));
+                }
+            }
+        }
+    }
+
+    @Override 
+    public String getName() throws SchemaException {
+        String schemaName = (String) jsonSchemaMap.get(TITLE_KEY);
+        if (schemaName != null) {
+            return schemaName;
+        } else {
+            throw new SchemaException("Invalid WSO2 Data Mapper JSON input schema, schema name not found.");
+        }
+    }
+
+    @Override 
+    public String getElementTypeByName(List<SchemaElement> elementStack)
+            throws InvalidPayloadException, SchemaException {
+        Map<String, Object> schema = jsonSchemaMap;
+        String elementType = null;
+        boolean elementFound;
+        for (SchemaElement element : elementStack) {
+            elementFound = false;
+            String elementName = element.getElementName();
+            if (elementName.equals(getName())) {
+                schema = (Map<String, Object>) jsonSchemaMap.get(PROPERTIES_KEY);
+                elementType = (String) jsonSchemaMap.get(TYPE_KEY);
+                if (ARRAY_ELEMENT_TYPE.equals(elementType)) {
+                    setCurrentArrayType(schema, elementName);
+                }
+                elementFound = true;
+            } else if (schema.containsKey(elementName)) {
+                elementType = (String) ((Map<String, Object>) schema.get(elementName)).get(TYPE_KEY);
+                if (ARRAY_ELEMENT_TYPE.equals(elementType)) {
+                    setCurrentArrayType(schema, elementName);
+                    schema = getSchemaItems((Map<String, Object>) schema.get(elementName));
+                    schema = getSchemaProperties(schema);
+                } else {
+                    schema = getSchemaProperties((Map<String, Object>) schema.get(elementName));
+                }
+                elementFound = true;
+            }
+            if (!elementFound) {
+                elementType = NULL_ELEMENT_TYPE;
+                log.warn("Element name not found : " + elementName);
+            }
+        }
+        return elementType;
+    }
+
+    private void setCurrentArrayType(Map<String, Object> schema, String elementName) {
```

Comment:
out of curiosity why is it Map<String, Object> instead of being Map<String, T extends someType> ?